File: MCPC18/D.py
# Author: John Qian

from collections import defaultdict


n, m = map(int, input().strip().split())

flights = []
for i in range(m):
    flights.append([int(a) for a in input().split()] + [i])

outfrom = defaultdict(list)
for flight in flights:
    outfrom[flight[0]].append(flight)

def frust(fi, F={}):
    if fi not in F:
        f = flights[fi]
        if f[1] == n:
            F[fi] = 0
        else:
            F[fi] = min((f2[2] - f[3])**2 + frust(f2[4]) for f2 in outfrom[f[1]] if f2[2] > f[3])
    return F[fi]

print(min(f[2]**2 + frust(f[4]) for f in outfrom[1]))

File: MCPC18/D.cpp
// Author: John Qian

#include <vector>
#include <cstdio>
#include <string.h>
#include <algorithm>

using namespace std;

typedef struct flight {
  int from;
  int to;
  int dep;
  int arr;

  bool operator<(const flight &other) const {
      return arr < other.arr;
  }
} flight;

flight flights[200010];
vector<int> outfrom[200010];
int main() {
    int n, m;
    scanf("%d %d", &n, &m);

    memset(flights, 0, sizeof(flights));
    int f, t, d, a;
    for (int i = 0; i < m; i++) {
        scanf("%d %d %d %d", &flights[i].from, &flights[i].to, &flights[i].dep, &flights[i].arr);
    }

    sort(flights, flights+m);

    for (int i = 0; i < m; i++) {
      outfrom[flights[i].from].push_back(i);
    }

    printf("%d\n", flights[0].from);
    printf("%d\n", outfrom[0][0]);
}
